examples/gcs-to-bq-serverless-services/Readme.md

```diff
@@ -0,0 +1,113 @@
+      --rules="all"
+      ````
+
+  - **Private Google Access:** The subnet must have [Private Google Access](https://cloud.google.com/vpc/docs/configure-private-google-access) enabled.
+      - External network access. Drivers and executors have internal IP addresses. You can set up [Cloud NAT](https://cloud.google.com/nat/docs/overview) to allow outbound traffic using internal IPs on your VPC network.
+
+- **Step 4:** Create necessary GCP resources required by Serverless Spark
+    - **Create BQ Dataset** Create a dataset to load GCS files. 
+        ```
+        DATASET_NAME=<<dataset_name>>
+        bq --location=US mk -d \
+            ${DATASET_NAME}
+        ```
+
+    -  **Create BQ table** Create a table using the schema in `schema/schema.json`
+        ```
+        TABLE_NAME=<<table_name>>
+        bq mk --table ${PROJECT_ID}:${DATASET_NAME}.${TABLE_NAME} \
+            ./schema/schema.json
+        ```
+
+    - **Create service account** Create service acccount used run the service account. We also create the permission required to read from GCS bucket, write to BigQuery table and publish error message in deadletter queue. The service account is used to run the serverless spark, so it needs dataproc worker role as well.
+  
+        ```
+        SERVICE_ACCOUNT_ID="gcs-to-bq-sa"
+        gcloud iam service-accounts create ${SERVICE_ACCOUNT_ID} \
+            --description="GCS to BQ service account for Serverless Spark" \
+            --display-name="GCS2BQ-SA"
+        
+        roles=("roles/dataproc.worker" "roles/bigquery.dataEditor" "roles/bigquery.jobUser" "roles/storage.objectViewer" "roles/pubsub.publisher")
+        for role in ${roles[@]}; do
+            gcloud projects add-iam-policy-binding ${PROJECT_ID} \
+                --member="serviceAccount:${SERVICE_ACCOUNT_ID}@${PROJECT_ID}.iam.gserviceaccount.com" \
+                --role="$role"
+        done
+        ```
+    - **Create BQ temp Bucket** GCS to BigQuery requires a temporary bucket. Lets create a temporary bucket
+        ```
+        GCS_TEMP_BUCKET=<<temp_bucket>>
+        gsutil mb gs://${GCS_TEMP_BUCKET}
+        ```
+    - **Create Deadletter Topic and Subscription** Lets create a dead letter topic and subscription
+
+        ```
+        ERROR_TOPIC=err_gcs2bq_${GCS_BUCKET_NAME}
+        gcloud pubsub topics create $ERROR_TOPIC
+        gcloud pubsub subscriptions create err_sub_${GCS_BUCKET_NAME}} \
+        --topic=${ERROR_TOPIC}
+        ```
+
+    Once all resources are create, please change the varaibles value () in `trigger-serverless-spark-fxn/main.py` from line 25 to 29
+
+        ```
+        bq_temp_bucket = <<GCS_TEMP_BUCKET>>
+        gcs_artifact_rep = <<GCS_ARTIFACT_REPO>>
+        dataset= <<DATASET_NAME>>
+        bq_table = <<TABLE_NAME>>
+        error_topic=<<ERROR_TOPIC>>
+        ```
+
+- **Step 5:** The cloud function is triggered one the object is copies to bucket.  The cloud function triggers the Servereless spark
```

Comment:
The cloud function is triggered **once** the object is **copied** to bucket.

Reply:
> 

Thanks @iht for the review.

Agreed with your point, for the pipeline receives large number of small files, the example is not performant and cost effective. The example was meant for larger files (example daily extract from sources systems). I will make this explicit in Readme.
For exception handling, The example spark is simple (toy example), hence for any exception it is Publish the error to Pub/Sub and stops. I am also writing errors from cloud function to Pub/Sub. Not by default, but cloud functions are tried multiple times(and hence the spark), so yes multiple and transient error might go be written in Pub/Sub. I will like the idea of making sure the DLQ having only single entry failed jobs, so pipelines can be retired with less human intervention. Please share if you have some examples/patterns, I will incorporate them. Meanwhile I will also make it clear in Readme.

The solution is missing some unit tests, I will add it.